### PR TITLE
Issue #404: Add security mailing list.

### PR DESCRIPTION
**Pull request by serundeputy**

### CODE REVIEW

**Comment by serundeputy** [comment]
I've committed the `dev-active` config files for the new `field_security_mailing_list` field, but now I'm questioning that what should really be committed is the `live-active` files and not the dev ones?

File: config/dev-active/field.instance.user.user.field_security_mailing_list.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
+    "_config_name": "field.instance.user.user.field_security_mailing_list",
+    "field_name": "field_security_mailing_list",
+    "entity_type": "user",
+    "bundle": "user",
+    "label": "Receive BackdropCMS.org security announcements for core and contrib modules",
```
Comment (jenlampton) [inline config/dev-active/field.instance.user.user.field_security_mailing_list.json]: 👍

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  // Check if we have sent mail for this node already.
+  $sent = _borg_mailers_check_if_already_sent($node->nid);
+
+  if (!$sent) {
+    // Check for node type is a project_release or core_release and status = 1.
+    if (($node->type == 'sa') && $node->status == 1) {
+      $language = language_default();
+      $params = array('node' => $node);
+      $to = _borg_mailers_security_subscribers();
+      foreach ($to as $t) {
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: Can we use a better variable name than `$t`? What is this, an email address? How about `$email` or `$address`?

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+/**
+ * Implements hook_mail().
+ */
+function borg_mailers_security_mail($key, &$message, $params) {
+  if ($key == 'borg_mailers_security_email') {
+    // Just copy over the subject and body as provided.
+    $message['subject'] = $params['node']->title;
+    $body = field_get_items('node', $params['node'], 'body');
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: Why do we use `field_get_items()` for the body but not the other fields?

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module (hunk 2 above)
Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: updated to `$address`

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    $message['body'][] = backdrop_html_to_text(
+      '<p>' .
+        t('View online: !url', array('!url' => url('node/' . $params['node']->nid, array('absolute' => TRUE)))) .
+      '</p>' . $body[0]['value'] .
+      '<p>' . $params['node']->field_sa_versions['und'][0]['value'] . ' </p>' .
+      '<p>' . $params['node']->field_sa_solution['und'][0]['value'] . '</p>'
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: No `field_get_items()` here?

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+ * An associative array of email addresses keyed on uid.
+ */
+function _borg_mailers_security_subscribers() {
+  $sql = db_select('field_data_field_security_mailing_list', 'sl')
+    ->fields('sl', array('entity_id'))
+    ->condition('field_security_mailing_list_value', 1)
+    ->execute();
+
+  $to = [];
+  foreach ($sql as $s) {
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: we should use better variable names here too. `foreach ($result as $subscriber)`

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.install [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  );
+
+  return $schema;
+}
+
+/**
+ * Implements hook_install().
+ *
+ * Populate the borg_mailers_security table with entries for existing nodes so
+ * we don't send emails about eld content.
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.install]: eld => old

Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.install]: ahh; i love `eld` ;)

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.install [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+ */
+function borg_mailers_security_install() {
+  // Get the current Security Announcements (sa).
+  $sas = db_select('node', 'n')
+    ->fields('n', array('nid'))
+    ->condition('type', 'sa')
+    ->execute();
+
+  foreach ($sas as $sa) {
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.install]: Optional: Better variable names here too? `foreach ($sas as $sa) {` could be `foreach ($announcements as $sa) {` or something.

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module [hunk 8]
```diff
@@ -0,0 +1,121 @@
+<?php
+/**
+ * @file borg_mailers_security.module
+ * Send out Security Anouncement emails to users who have subscribed.
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: users -> people?

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * Users can check a box upon editing there user profile (/user when logged in)
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: Users -> People?

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+/**
+ * Implements hook_node_submit().
+ *
+ * @see _borg_mailers_record_mail_sent().
+ * @see _borg_mailers_security_subscribers().
+ * @see _borg_mailers_record_mail_sent().
+ */
+function borg_mailers_security_node_submit($node) {
+  // Save the node right quick so we can get an nid.
+  $node->save();
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: It seems that saving the node before the system is ready to save the node would be problematic. Maybe we need to use `hook_node_presave()` or some other hook where we already have a NID...

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+  // Check if we have sent mail for this node already.
+  $sent = _borg_mailers_check_if_already_sent($node->nid);
+
+  if (!$sent) {
+    // Check for node type is a project_release or core_release and status = 1.
```
Comment (jenlampton) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: Is this comment still accurate? I don't see these node types in the check...

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module (hunk 3 above)
Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: I've changed them all to `field_get_items` now; PR Is updated.

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module (hunk 5 above)
Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: updated.

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.install (hunk 6 above)
Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.install]: updated.

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.install (hunk 7 above)
Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.install]: updated.

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module (hunk 8 above)
Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: updated.

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module (hunk 9 above)
Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: updated.

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module (hunk 11 above)
Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: removed; it was no longer accurate.

File: www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module (hunk 10 above)
Comment (serundeputy) [inline www/modules/custom/borg_mailers/borg_mailers_security/borg_mailers_security.module]: I've added a worker function and called it from `hook_node_insert` and `hook_node_update`
@jenlampton I think we are close?

**Pull request merged**